pkg/tfgen/internal/unrec/comparer.go

```diff
@@ -0,0 +1,308 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unrec
+
+import (
+	"bytes"
+	"reflect"
+
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+)
+
+type comparer struct {
+	// Local type-ref comparisons are scoped to a package schema.
+	schema *pschema.PackageSpec
+}
+
+func (cmp *comparer) EqualTypeRefs(a, b tokens.Type) bool {
+	g := &generalizedComparer{schema: cmp.schema}
+	g.EqualXPropertyMaps = g.strictlyEqualXPropertyMaps
+	return g.EqualTypeRefs(a, b)
+}
+
+func (cmp *comparer) LessThanTypeRefs(a, b tokens.Type) bool {
+	return cmp.LessThanOrEqualTypeRefs(a, b) && !cmp.EqualTypeRefs(a, b)
+}
+
+// A type will be considered "less than" another type if both are locally defined object types and A defines a subset of
+// B's properties. This is useful to deal with property dropout during recursive type expansions.
+func (cmp *comparer) LessThanOrEqualTypeRefs(a, b tokens.Type) (eq bool) {
+	g := &generalizedComparer{schema: cmp.schema}
+	g.EqualXPropertyMaps = g.lessThanOrEqualXPropertyMaps
+	return g.EqualTypeRefs(a, b)
+}
+
+// Generalizing structural comparisons to specialize for A=B and A<=B separately.
+type generalizedComparer struct {
+	schema             *pschema.PackageSpec
+	EqualXPropertyMaps func(xPropertyMap, xPropertyMap) bool
+}
+
+func (cmp *generalizedComparer) lessThanOrEqualXPropertyMaps(a, b xPropertyMap) bool {
+	for aK, aP := range a {
+		bP, ok := b[aK]
+		// Every key in A should also be a key in B.
+		if !ok {
+			return false
+		}
+		if !cmp.EqualXProperties(aP, bP) {
+			return false
+		}
+	}
+	return true
+}
+
+func (cmp *generalizedComparer) strictlyEqualXPropertyMaps(a, b xPropertyMap) bool {
+	if a == nil || b == nil {
+		return a == nil && b == nil
+	}
+	if len(a) != len(b) {
+		return false
+	}
+	for k, av := range a {
+		bv, ok := b[k]
+		if !ok {
+			return false
+		}
+		if !cmp.EqualXProperties(av, bv) {
+			return false
+		}
+	}
+	return true
+}
+
+func (cmp *generalizedComparer) EqualTypeRefs(a, b tokens.Type) bool {
+	if a == b {
+		return true
+	}
+	aT, gotA := cmp.schema.Types[string(a)]
+	bT, gotB := cmp.schema.Types[string(b)]
+	if gotA && gotB {
+		return cmp.EqualComplexTypeSpecs(&aT, &bT)
+	}
+	return false
+}
+
+func (cmp *generalizedComparer) EqualRawRefs(a, b string) bool {
+	if a == b {
+		return true
+	}
+	aT, ok1 := parseLocalRef(a)
+	bT, ok2 := parseLocalRef(b)
+	if ok1 && ok2 {
+		return cmp.EqualTypeRefs(aT, bT)
+	}
+	return false
+}
+
+func (cmp *generalizedComparer) EqualXProperties(a, b xProperty) bool {
+	if a.IsRequired != b.IsRequired {
+		return false
+	}
+	if a.IsPlain != b.IsPlain {
+		return false
+	}
+	if !cmp.EqualPropertySpecs(&a.PropertySpec, &b.PropertySpec) {
+		return false
+	}
+	return true
+}
+
+func (cmp *generalizedComparer) EqualObjectTypeSpecs(a, b pschema.ObjectTypeSpec) bool {
+	if a.Type != b.Type {
+		return false
+	}
```

Comment:
Why don't we delegate to `cmp.EqualTypeSpecs` here?
```suggestion
	if !cmp.EqualTypeSpecs(a.Type, b.Type) {
		return false
	}
```

Reply:
a.Type is a string like "object":

```
	// Type must be "object" if this is an object type, or the underlying type for an enum.
	Type string `json:"type,omitempty" yaml:"type,omitempty"`
```

but

```
EqualTypeSpecs(a, b *pschema.TypeSpec) bool
```

I think it should stay as written.